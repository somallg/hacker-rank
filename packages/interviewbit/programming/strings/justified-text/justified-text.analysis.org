| Question                    | Answer |
|-----------------------------+--------|
| Yeu cau la gi ?             |        |
| Yeu cau dac biet gi khong ? |        |
| Y tuong gi khong ?          |        |
| Phan tich y tuong tren ?    |        |

* DONE Doc va phan tich yeu cau
  CLOSED: [2018-08-10 Fri 17:46]
  :LOGBOOK:
  CLOCK: [2018-08-10 Fri 17:35]--[2018-08-10 Fri 17:46] =>  0:11
  :END:
- Cho mot mang gom gom cac chu ~words~ va do dai ~L~, can giua cac tu (words) sao cho moi dong khong duoc qua do dai L
- Khi chon chu thi ta can chon nhieu nhat co the moi dong
- Moi dong khi can giua, cac chu deu nhat co the
- Neu nhu khoang cach khong the chia deu thi so luong khoang cach ben trai se nhieu hon ben phai
- Dong cuoi cung se duoc can trai va khong co dau cach the nao giua cac so
* DONE TDD: Setup test case
  CLOSED: [2018-08-10 Fri 17:51]
  :LOGBOOK:
  CLOCK: [2018-08-10 Fri 17:46]--[2018-08-10 Fri 17:51] =>  0:05
  :END:

* DONE Dua ra brute force solution
  CLOSED: [2018-08-10 Fri 17:58]
  :LOGBOOK:
  CLOCK: [2018-08-10 Fri 17:51]--[2018-08-10 Fri 17:58] =>  0:07
  :END:
Brute force solution co 2 buoc
- Buoc 1: lay cang nhieu words vao 1 dong nhat co the
  - Loop tung chu trong mang, tinh do dai cho moi dong
  - 2 tu se co do dai it nhat la len(word1) + len(word2) + 1 (space)
  - Khi loop den wordi neu len + len(wordi) + 1 <= L thi push wordi vao dong dang lap
  - Neu khong thi push wordi vao dong moi
  - Done
- Buoc 2: can chinh tinh tu trong mang
  - Duyet tung mang o buoc 1
  - Tinh so luong dau cach: nbSpaces = L / Sum(words),
  - neu nbSpaces la so tu nhien thi noi cac phan tu trong mang
  - neu khong phan du ra se duoc cho vao ben trai
  - den phan tu cuoi cung thi chi can trai
* DONE Toi uu vao so sanh cac thuat toan
  CLOSED: [2018-08-10 Fri 17:59]
  :LOGBOOK:
  CLOCK: [2018-08-10 Fri 17:59]--[2018-08-10 Fri 17:59] =>  0:00
  :END:
Chua nghi ra cach toi uu
* DONE Bat dau code thuat toan thich hop nhat
  CLOSED: [2018-08-10 Fri 18:14]
  :LOGBOOK:
  CLOCK: [2018-08-10 Fri 17:59]--[2018-08-10 Fri 18:14] =>  0:15
  :END:
Coding done
* DONE Test va fix bug neu co
  CLOSED: [2018-08-11 Sat 00:03]
  :LOGBOOK:
  CLOCK: [2018-08-10 Fri 18:14]--[2018-08-11 Sat 00:03] =>  5:49
  :END:

#+BEGIN: clocktable :scope file :maxlevel 2
#+CAPTION: Clock summary at [2018-08-07 Tue 22:10]
| Headline     | Time   |
|--------------+--------|
| *Total time* | *0:00* |
#+END:
