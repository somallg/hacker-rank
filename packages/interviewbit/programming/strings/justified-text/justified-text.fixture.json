{
  "name": "JustifiedText Fixture",
  "testCategories": [
    {
      "name": "Example Tests",
      "testCases": [
        {
          "name": "Example 1",
          "input": [
            ["This", "is", "an", "example", "of", "text", "justification."],
            16
          ],
          "output": ["This    is    an", "example  of text", "justification.  "]
        },
        {
          "name": "Example 2",
          "input": [
            [
              "am",
              "fasgoprn",
              "lvqsrjylg",
              "rzuslwan",
              "xlaui",
              "tnzegzuzn",
              "kuiwdc",
              "fofjkkkm",
              "ssqjig",
              "tcmejefj",
              "uixgzm",
              "lyuxeaxsg",
              "iqiyip",
              "msv",
              "uurcazjc",
              "earsrvrq",
              "qlq",
              "lxrtzkjpg",
              "jkxymjus",
              "mvornwza",
              "zty",
              "q",
              "nsecqphjy"
            ],
            14
          ],
          "output": [
            "am    fasgoprn",
            "lvqsrjylg     ",
            "rzuslwan xlaui",
            "tnzegzuzn     ",
            "kuiwdc        ",
            "fofjkkkm      ",
            "ssqjig        ",
            "tcmejefj      ",
            "uixgzm        ",
            "lyuxeaxsg     ",
            "iqiyip     msv",
            "uurcazjc      ",
            "earsrvrq   qlq",
            "lxrtzkjpg     ",
            "jkxymjus      ",
            "mvornwza zty q",
            "nsecqphjy     "
          ]
        }
      ]
    },
    {
      "name": "Correctness Tests",
      "testCases": [
        {
          "name": "Sample 1",
          "input": [["ab", "cd"], 2],
          "output": ["ab", "cd"]
        },
        {
          "name": "Sample 2",
          "input": [["What", "must", "be", "shall", "be."], 12],
          "output": ["What must be", "shall be.   "]
        },
        {
          "name": "Extreme 1",
          "input": [[""], 10],
          "output": []
        }
      ]
    },
    {
      "name": "Performance Tests",
      "testCases": [
        {
          "name": "Medium 1",
          "inputSize": 100
        },
        {
          "name": "Medium 2",
          "inputSize": 1e3
        },
        {
          "name": "Large 1",
          "inputSize": 10e3
        },
        {
          "name": "Large 2",
          "inputSize": 50e3
        },
        {
          "name": "Large 3",
          "inputSize": 100e3
        }
      ]
    }
  ]
}
