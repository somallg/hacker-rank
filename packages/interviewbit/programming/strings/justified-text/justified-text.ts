/**
 * JustifiedText
 */

const SPC = ' ';

function packWords(words: string[], len: number): string[][] {
  const result: string[][] = [[]];

  let index = 0;
  let l = -1;

  words.forEach(w => {
    if (l + w.length + 1 <= len) {
      result[index].push(w);
      l += w.length + 1;
    } else {
      index += 1;
      result[index] = [w];
      l = w.length;
    }
  });

  return result;
}

function join(words: string[], paddingLeft: number, nbSpaces: number) {
  const l = words.length;
  return words.reduce((acc, w, index) => {
    if (index === l - 1) {
      return acc + w;
    }

    return acc + w + SPC.repeat(nbSpaces + (paddingLeft-- > 0 ? 1 : 0));
  }, '');
}

function alignLeft(words: string[], wordsLen: number, len: number) {
  if (wordsLen === 0) {
    return '';
  }
  return words.join(SPC) + SPC.repeat(len - wordsLen - words.length + 1);
}

function justifiedText(words: string[], len: number): string[] {
  const lines: string[][] = packWords(words, len);
  const result: string[] = [];

  lines.forEach((line, index) => {
    const lineLen = line.reduce((acc, word) => acc + word.length, 0);
    if (line.length === 1 || index === lines.length - 1) {
      result.push(alignLeft(line, lineLen, len));
    } else {
      const nbSpaces = Math.floor((len - lineLen) / (line.length - 1));
      const paddingLeft = (len - lineLen) % (line.length - 1);
      result.push(join(line, paddingLeft, nbSpaces));
    }
  });

  return result.filter(w => w.length > 0);
}

export { justifiedText };
