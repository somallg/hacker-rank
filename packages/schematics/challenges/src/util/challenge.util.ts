import { strings } from '@angular-devkit/core';

interface SwitchOptions {
  [key: string]: string;
}

function getChallengeName(path: string): string {
  const [, challenge] = path.split('/');

  return challenge;
}

function getFunctionName(challengeName: string, problem: string): string {
  // set default name to camelCase
  const camelizedProblemName = strings.camelize(problem);
  const classifiedProblemName = strings.classify(problem);

  return (
    ({
      codility: 'solution',
      hackerearth: `solve${classifiedProblemName}`,
      hackerrank: `solve${classifiedProblemName}`,
      interviewbit: camelizedProblemName,
      leetcode: camelizedProblemName,
      spoj: camelizedProblemName
    } as SwitchOptions)[challengeName] || camelizedProblemName
  );
}

function getUrl(challengeName: string, problem: string) {
  return (
    ({
      hackerrank: ` @url https://www.hackerrank.com/challenges/${problem}/problem`
    } as SwitchOptions)[challengeName] || strings.classify(problem)
  );
}

export { getChallengeName, getFunctionName, getUrl };
