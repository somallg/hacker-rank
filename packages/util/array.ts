/*
 * Utils function which generate an array of size length
 * Depend on the value of random, if random is true
 * then it generata an array where each element of the array
 * in range [startFrom, ...maxExclusive - 1].
 * If random is false then it return an array [0...0] of size length
 */

export { generateArray, generateStringArray, generateMatrix, zip };

function generateArray(
  length: number,
  maxExclusive: number = length,
  startFrom: number = 0,
  random: boolean = true
): number[] {
  const res: number[] = [];

  for (let i: number = 0; i < length; i = i + 1) {
    // tslint:disable:insecure-random
    res.push(random ? Math.floor(Math.random() * maxExclusive + startFrom) : 0);
  }

  return res;
}

function generateStringArray(length: number): string[] {
  return generateArray(length).map(String);
}

function generateMatrix(
  length: number,
  maxExclusive: number = length
): number[][] {
  const result: number[][] = [];

  for (let i: number = 0; i < length; i = i + 1) {
    result[i] = generateArray(length, maxExclusive);
  }

  return result;
}

function zip<T>(arrayA: T[], arrayB: T[]): T[][] {
  return arrayA.map((a: T, index: number) => [a, arrayB[index]]);
}
