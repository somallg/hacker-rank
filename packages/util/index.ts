export * from './array';
export * from './memorized';
export * from './string';
export * from './unit-test.util';
