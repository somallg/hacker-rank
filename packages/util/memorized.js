function memorized(fn) {
  var cached = {};

  return function(n) {
    if (cached[n] === undefined) {
      cached[n] = fn(n);
    }

    return cached[n];
  };
}

module.exports = {
  memorized
};
