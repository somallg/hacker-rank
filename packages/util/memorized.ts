type Immutable = string | number;

type MemoFn<InputType, OutputType> = (x: InputType) => OutputType;

function id<InputType>(x: InputType): Immutable {
  if (typeof x === 'number') {
    return x;
  } else if (typeof x === 'string') {
    return x;
  }

  return JSON.stringify(x);
}

function memorized<InputType, OutputType>(
  fn: MemoFn<InputType, OutputType>,
  keyFn = id
): MemoFn<InputType, OutputType> {
  const memo = new Map<Immutable, OutputType>();

  return (x: InputType): OutputType => {
    let cachedValue = memo.get(keyFn(x)) as OutputType;
    if (cachedValue === undefined) {
      cachedValue = fn(x);
      memo.set(keyFn(x), cachedValue);
    }

    return cachedValue;
  };
}

export { memorized };
