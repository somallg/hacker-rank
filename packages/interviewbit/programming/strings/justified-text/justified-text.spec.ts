/**
 * Test specs for justified-text problem
 * JustifiedText
 */

import { generateArray } from '@challenges/util';

import { justifiedText } from './justified-text';

describe('Interviewbit - JustifiedText', () => {
  describe('justifiedText', () => {
    describe('Example Tests', () => {
      it('should return correctly for Example 1 input', () =>
        expect(
          justifiedText(
            ['This', 'is', 'an', 'example', 'of', 'text', 'justification.'],
            16
          )
        ).toMatchSnapshot());

      it('should return correctly for Example 2 input', () =>
        expect(
          justifiedText(
            [
              'am',
              'fasgoprn',
              'lvqsrjylg',
              'rzuslwan',
              'xlaui',
              'tnzegzuzn',
              'kuiwdc',
              'fofjkkkm',
              'ssqjig',
              'tcmejefj',
              'uixgzm',
              'lyuxeaxsg',
              'iqiyip',
              'msv',
              'uurcazjc',
              'earsrvrq',
              'qlq',
              'lxrtzkjpg',
              'jkxymjus',
              'mvornwza',
              'zty',
              'q',
              'nsecqphjy'
            ],
            14
          )
        ).toMatchSnapshot());
    });

    describe('Correctness Tests', () => {
      it('should return correctly for Sample 1 input', () =>
        expect(justifiedText(['ab', 'cd'], 2)).toMatchSnapshot());

      it('should return correctly for Sample 2 input', () =>
        expect(
          justifiedText(['What', 'must', 'be', 'shall', 'be.'], 12)
        ).toMatchSnapshot());

      it('should return correctly for Extreme 1 input', () =>
        expect(justifiedText([''], 10)).toMatchSnapshot());
    });

    describe('Performance Tests', () => {
      it('should run for Medium 1 input', () =>
        expect(
          justifiedText(generateArray(100).map(String), 10)
        ).toBeDefined());

      it('should run for Medium 2 input', () =>
        expect(
          justifiedText(generateArray(1e3).map(String), 10)
        ).toBeDefined());

      it('should run for Large 1 input', () =>
        expect(
          justifiedText(generateArray(10e3).map(String), 100)
        ).toBeDefined());

      it('should run for Large 2 input', () =>
        expect(
          justifiedText(generateArray(50e3).map(String), 100)
        ).toBeDefined());

      it('should run for Large 3 input', () =>
        expect(
          justifiedText(generateArray(100e3).map(String), 100)
        ).toBeDefined());
    });
  });
});
