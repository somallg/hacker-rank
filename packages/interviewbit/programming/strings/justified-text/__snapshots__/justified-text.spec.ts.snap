// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interviewbit - JustifiedText justifiedText Correctness Tests should return correctly for Example 1 input 1`] = `
Array [
  "ab",
  "cd",
]
`;

exports[`Interviewbit - JustifiedText justifiedText Correctness Tests should return correctly for Example 2 input 1`] = `
Array [
  "What must be",
  "shall be.   ",
]
`;

exports[`Interviewbit - JustifiedText justifiedText Correctness Tests should return correctly for Extreme 1 input 1`] = `Array []`;

exports[`Interviewbit - JustifiedText justifiedText Example Tests should return correctly for Example 1 input 1`] = `
Array [
  "This    is    an",
  "example  of text",
  "justification.  ",
]
`;

exports[`Interviewbit - JustifiedText justifiedText Example Tests should return correctly for Example 2 input 1`] = `
Array [
  "am    fasgoprn",
  "lvqsrjylg     ",
  "rzuslwan xlaui",
  "tnzegzuzn     ",
  "kuiwdc        ",
  "fofjkkkm      ",
  "ssqjig        ",
  "tcmejefj      ",
  "uixgzm        ",
  "lyuxeaxsg     ",
  "iqiyip     msv",
  "uurcazjc      ",
  "earsrvrq   qlq",
  "lxrtzkjpg     ",
  "jkxymjus      ",
  "mvornwza zty q",
  "nsecqphjy     ",
]
`;
